### ISSUE TITLE
Output Mismatch of torch.mul when applying equivalent torch.transpose on GPU

### ISSUE THREAD

**Issue by Azyka** [NVIDIA/TensorRT#3447]
## Description
Adding multiple torch.Tensor.transpose nodes before and after a node should get the same output as the node without applying transpose, like `transpose(torch.mul(transpose(i1), transpose(i2))) = torch.mul(i1, i2)`. However, the output of `torch.mul` turns out to be changed when applying equivalent torch.transpose, causing mismatch of model outputs.
Original model:
<img width="200" alt="image" src="https://github.com/NVIDIA/TensorRT/assets/74590664/23286c2f-e2b4-4276-8b78-d1217201fd4c">

Transpose model:
<img width="234" alt="image" src="https://github.com/NVIDIA/TensorRT/assets/74590664/303900c3-9aa8-4cd9-a60d-b40eeff69cb6">

## Environment

**TensorRT Version**: 8.6.1.post1

**NVIDIA GPU**: RTX 1660

**NVIDIA Driver Version**: 525.147.05

**CUDA Version**: 12.0

**CUDNN Version**: 8.9.4.25

Operating System: Ubuntu 22.04.3 LTS (x86_64)

Python Version (if applicable): 3.10.12

Tensorflow Version (if applicable): 2.13.0

PyTorch Version (if applicable): 2.1.0+cu118

## Relevant Files

**Model link**: [models.zip](https://github.com/NVIDIA/TensorRT/files/13334800/models.zip)
Input data file: [input_data.zip](https://github.com/NVIDIA/TensorRT/files/13334807/input_data.zip)

## Steps To Reproduce

**Script**:
```python
from dataclasses import dataclass
from numpy import testing
import numpy as np
import torch
import tensorrt as trt
import pycuda.driver as cuda
from pycuda.driver import DeviceAllocation
import pickle


@dataclass
class HostDeviceMem:
    host: np.ndarray
    device: DeviceAllocation


class ONNXClassifierWrapper():
    def __init__(self, engine):
        self.engine = engine

    def allocate_memory(self):
        engine = self.engine
        inputs = []
        outputs = []
        bindings = []
        stream = cuda.Stream()
        onames = []
        name2idx = {}
        for idx, binding in enumerate(engine):
            name2idx[binding] = idx
            size = trt.volume(engine.get_binding_shape(binding)) * engine.max_batch_size
            dtype = trt.nptype(engine.get_binding_dtype(binding))
            # Allocate host and device buffers
            host_mem = cuda.pagelocked_empty(size, dtype)
            device_mem = cuda.mem_alloc(host_mem.nbytes)
            # Append the device buffer to device bindings.
            bindings.append(int(device_mem))
            # Append to the appropriate list.
            if engine.binding_is_input(binding):
                inputs.append(HostDeviceMem(host_mem, device_mem))
            else:
                outputs.append(HostDeviceMem(host_mem, device_mem))
                onames.append(binding)
        return inputs, outputs, bindings, stream, onames, name2idx

    def predict(self, inputs):  # result gets copied into output
        (
            trt_inputs,
            trt_outputs,
            trt_bindings,
            stream,
            onames,
            name2idx,
        ) = self.allocate_memory()
        context = self.engine.create_execution_context()
        # print(name2idx)
        # print(inputs.keys())
        for iname in inputs:
            np.copyto(
                trt_inputs[name2idx[iname]].host,
                inputs[iname]
                .astype(trt.nptype(self.engine.get_binding_dtype(iname)))
                .ravel(),
            )

        [cuda.memcpy_htod_async(inp.device, inp.host, stream) for inp in trt_inputs]
        context.execute_async_v2(bindings=trt_bindings, stream_handle=stream.handle)
        [cuda.memcpy_dtoh_async(out.host, out.device, stream) for out in trt_outputs]
        stream.synchronize()
        trt_outputs = [out.host for out in trt_outputs]
        return {
            n: v.reshape(self.engine.get_binding_shape(n))
            for n, v in zip(onames, trt_outputs)
        }


def convert_onnx_to_engine(onnx_filename):
    logger = trt.Logger(trt.Logger.WARNING)
    with trt.Builder(logger) as builder, builder.create_network(
            1 << (int)(trt.NetworkDefinitionCreationFlag.EXPLICIT_BATCH)) as network, \
            trt.OnnxParser(network, logger) as parser:
        # builder.max_workspace_size = max_workspace_size
        # builder.fp16_mode = fp16_mode
        # builder.max_batch_size = max_batch_size

        config = builder.create_builder_config()
        config.set_memory_pool_limit(trt.MemoryPoolType.WORKSPACE, 2 << 30)

        with open(onnx_filename, 'rb') as model:
            if not parser.parse(model.read()):
                for error in range(parser.num_errors):
                    print(parser.get_error(error))

        engine_bytes = builder.build_serialized_network(network, config)

        return trt.Runtime(trt.Logger(trt.Logger.WARNING)).deserialize_cuda_engine(
            engine_bytes
        )

import pycuda.autoinit

p0 = torch.nn.Parameter(torch.empty([30, 39, 1], dtype=torch.int32), requires_grad=False).to('cuda')

class Model0(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.v4_0 = p0

    def forward(self, *args):
        _args = args
        v4_0 = self.v4_0
        getitem_2 = _args[0]
        mul = torch.mul(getitem_2, v4_0)
        return (mul)

model_0 = Model0()
output_names_0 = ['v2_0']
input_dict_0 = pickle.load(open('0.pickle', 'rb'))
inputs_0 = tuple(torch.from_numpy(v).to('cuda') for _, v in input_dict_0.items())
torch.onnx.export(model_0, inputs_0, '0.onnx', verbose=False, input_names=['v0_0'], output_names=output_names_0, opset_version=14, do_constant_folding=False)

class Model1(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.v2_0 = p0

    def forward(self, *args):
        _args = args
        v2_0 = self.v2_0
        getitem = _args[0]
        transpose = getitem.transpose(0, 1)
        transpose_1 = v2_0.transpose(1, 0)
        mul = torch.mul(transpose, transpose_1)
        transpose_2 = mul.transpose(1, 0)
        return (transpose_2)

model_1 = Model1()
output_names_1 = ['v2_0']
input_dict_1 = pickle.load(open('0.pickle', 'rb'))
inputs_1 = tuple(torch.from_numpy(v).to('cuda') for _, v in input_dict_1.items())
torch.onnx.export(model_1, inputs_1, '1.onnx', verbose=False, input_names=['v0_0'], output_names=output_names_1, opset_version=14, do_constant_folding=False)

engine_0 = convert_onnx_to_engine('0.onnx')
wrapper_0 = ONNXClassifierWrapper(engine_0)
output_0 = wrapper_0.predict(input_dict_0)

engine_1 = convert_onnx_to_engine('1.onnx')
wrapper_1 = ONNXClassifierWrapper(engine_1)
output_1 = wrapper_1.predict(input_dict_1)
output_name_dict = {'v2_0': 'v2_0'}

print('=========================')
try:
    for tensor_name_0, tensor_name_1 in output_name_dict.items():
        testing.assert_allclose(output_0[tensor_name_0], output_1[tensor_name_1], rtol=1, err_msg=f'at {tensor_name_0}, {tensor_name_1}')
    print("tensorRT does not trigger assertion")
except AssertionError as e:
    print("tensorRT triggers assertion")
    print(e)
print('=========================')
```

Steps to repro:
1. Download the input data file and put it at the same dir of the script.
2. Run the script

Output assertion:
```
=========================
tensorRT triggers assertion

Not equal to tolerance rtol=1, atol=0
at v2_0, v5_0
Mismatched elements: 449 / 1170 (38.4%)
Max absolute difference: 2147464543
Max relative difference: 66216049.
 x: array([[[      9500],
        [      1024],
        [ 352321536],...
 y: array([[[     9500],
        [  8359936],
        [  6796800],...
=========================
```

**Have you tried [the latest release](https://developer.nvidia.com/tensorrt)?**: No

**Can this model run on other frameworks?** For example run ONNX model with ONNXRuntime (`polygraphy run <model.onnx> --onnxrt`): Yes

**Comment by zerollzeng** [comment]
I can reproduce this in TRT 8.6 but looks like it's been fix in TRT 9.1. Here is much simpler reproduce code with [Polygraphy](https://github.com/NVIDIA/TensorRT/tree/main/tools/Polygraphy)
```
import numpy as np
import pickle
from polygraphy.backend.trt import CreateConfig, EngineFromNetwork, NetworkFromOnnxPath, SaveEngine, TrtRunner

def main():
    engine0 = EngineFromNetwork(NetworkFromOnnxPath("0.onnx"))
    engine1 = EngineFromNetwork(NetworkFromOnnxPath("1.onnx"))

    data = pickle.load(open('0.pickle', 'rb'))
    input_data = np.array(data["v0_0"], dtype=np.int32)

    with TrtRunner(engine0) as runner0:
        outputs0 = runner0.infer(feed_dict={"v0_0": input_data})["v2_0"]
    with TrtRunner(engine1) as runner1:
        outputs1 = runner1.infer(feed_dict={"v0_0": input_data})["v5_0"]

    assert np.array_equal(outputs0, outputs1)
    print("Test pass")

if __name__ == "__main__":
    main()

```
Result
```
$ python3 compare.py 
[W] CUDA lazy loading is not enabled. Enabling it can significantly reduce device memory usage and speed up TensorRT initialization. See "Lazy Loading" section of CUDA documentation https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#lazy-loading
[I]     Configuring with profiles: [Profile().add('v0_0', min=[30, 39, 1], opt=[30, 39, 1], max=[30, 39, 1])]
[I] Building engine with configuration:
    Flags                  | []
    Engine Capability      | EngineCapability.DEFAULT
    Memory Pools           | [WORKSPACE: 22551.38 MiB, TACTIC_DRAM: 22551.38 MiB]
    Tactic Sources         | [CUBLAS, CUDNN, EDGE_MASK_CONVOLUTIONS, JIT_CONVOLUTIONS]
    Profiling Verbosity    | ProfilingVerbosity.DETAILED
    Preview Features       | [FASTER_DYNAMIC_SHAPES_0805, DISABLE_EXTERNAL_TACTIC_SOURCES_FOR_CORE_0805]
[I] Finished engine building in 0.148 seconds
[I]     Configuring with profiles: [Profile().add('v0_0', min=[30, 39, 1], opt=[30, 39, 1], max=[30, 39, 1])]
[I] Building engine with configuration:
    Flags                  | []
    Engine Capability      | EngineCapability.DEFAULT
    Memory Pools           | [WORKSPACE: 22551.38 MiB, TACTIC_DRAM: 22551.38 MiB]
    Tactic Sources         | [CUBLAS, CUDNN, EDGE_MASK_CONVOLUTIONS, JIT_CONVOLUTIONS]
    Profiling Verbosity    | ProfilingVerbosity.DETAILED
    Preview Features       | [FASTER_DYNAMIC_SHAPES_0805, DISABLE_EXTERNAL_TACTIC_SOURCES_FOR_CORE_0805]
[I] Finished engine building in 0.143 seconds
Test pass

```

**Comment by zerollzeng** [comment]
I saw you file several similar issues, would be great if you can try TRT 9.1 first to see if they still reproduce.

**Comment by Azyka** [comment]
> I can reproduce this in TRT 8.6 but looks like it's been fix in TRT 9.1. Here is much simpler reproduce code with [Polygraphy](https://github.com/NVIDIA/TensorRT/tree/main/tools/Polygraphy)
> 
> ```
> import numpy as np
> import pickle
> from polygraphy.backend.trt import CreateConfig, EngineFromNetwork, NetworkFromOnnxPath, SaveEngine, TrtRunner
> 
> def main():
>     engine0 = EngineFromNetwork(NetworkFromOnnxPath("0.onnx"))
>     engine1 = EngineFromNetwork(NetworkFromOnnxPath("1.onnx"))
> ```

Sure, and I will test the reproducibility when I submit new issues.
